#
# Fixes theme path parsing for image generation.
#
# If the theme's structure nests its resources and WordPress theme files in
# separate sub-directories of the theme's root directory (such is the case for
# Timber's Starter Theme v2), the `ImageHelper::theme_url_to_dir()` method would
# fail to correctly resolve the path to an image.
#
# This patch is a subset of the changeset of timber/timber#2458.
#
# See:
# - https://github.com/timber/timber/pull/2458
# - https://github.com/timber/starter-theme/tree/2.x
#
diff --git a/src/ImageHelper.php b/src/ImageHelper.php
index 48f99ed4..f55a0967 100644
--- a/src/ImageHelper.php
+++ b/src/ImageHelper.php
@@ -577,10 +577,10 @@ class ImageHelper
      */
     public static function theme_url_to_dir($src)
     {
-        $site_root = trailingslashit(get_theme_root_uri()) . get_stylesheet();
+        $site_root = trailingslashit(get_theme_root_uri()) . ThemeHelper::get_stylesheet();
         $tmp = str_replace($site_root, '', $src);
         //$tmp = trailingslashit(get_theme_root()).get_stylesheet().$tmp;
-        $tmp = get_stylesheet_directory() . $tmp;
+        $tmp = ThemeHelper::get_stylesheet_directory() . $tmp;
         if (realpath($tmp)) {
             return realpath($tmp);
         }
diff --git a/src/ThemeHelper.php b/src/ThemeHelper.php
new file mode 100644
index 00000000..2b676e03
--- /dev/null
+++ b/src/ThemeHelper.php
@@ -0,0 +1,102 @@
+<?php
+
+namespace Timber;
+
+/**
+ * Class ThemeHelper
+ *
+ * Helper class to work with current & parent theme's names, paths & directories.
+ *
+ * All class methods behave just like their corresponding WordPress functions,
+ * with the noticeable difference that they won't return what comes after "/" in theme's name.
+ *
+ * For example, consider a theme with the following tree structure (2.x default child theme structure):
+ * - my-theme-name/
+ * -> images/
+ * --> foo.jpg
+ * -> theme/
+ * --> functions.php
+ * --> style.css
+ *
+ * Theme template in style.css is my-theme-name/theme
+ *
+ * Regular get_stylesheet_directory_uri() will output something like:
+ * https://example.com/my-theme-name/theme
+ *
+ * When it should return the root uri of your theme, regardless of tree structure:
+ * https://example.com/my-theme-name
+ *
+ * @api
+ */
+class ThemeHelper
+{
+	/**
+	 * Retrieves and parses name of the current stylesheet.
+	 *
+	 * @see get_stylesheet()
+	 *
+	 * @return string Clean stylesheet name.
+	 */
+	public static function get_stylesheet() {
+		list($stylesheet) = explode('/', get_stylesheet());
+
+		return $stylesheet;
+	}
+
+	/**
+	 * Retrieves and parses stylesheet directory path for current theme.
+	 *
+	 * @see get_stylesheet_directory()
+	 *
+	 * @return string Clean path to current theme's stylesheet directory.
+	 */
+	public static function get_stylesheet_directory() {
+		return str_replace(get_stylesheet(), self::get_stylesheet(), get_stylesheet_directory());
+	}
+
+	/**
+	 * Retrieves and parses stylesheet directory URI for current theme.
+	 *
+	 * @see get_stylesheet_directory_uri()
+	 *
+	 * @return string Clean URI to current theme's stylesheet directory.
+	 */
+	public static function get_stylesheet_directory_uri() {
+		return str_replace(get_stylesheet(), self::get_stylesheet(), get_stylesheet_directory_uri());
+	}
+
+	/**
+	 * Retrieves and parses name of the current theme.
+	 *
+	 * @see get_template()
+	 *
+	 * @return string Clean template name.
+	 */
+	public static function get_template() {
+		list($template) = explode('/', get_template());
+
+		return $template;
+	}
+
+	/**
+	 * Retrieves and parses template directory path for current theme.
+	 *
+	 * @see get_template_directory()
+	 *
+	 * @return string Clean path to current theme's template directory.
+	 */
+	public static function get_template_directory() {
+		return str_replace(get_template(), self::get_template(), get_template_directory());
+	}
+
+	/**
+	 * Retrieves and parses template directory URI for current theme.
+	 *
+	 * @see get_template_directory_uri()
+	 *
+	 * @return string Clean URI to current theme's template directory.
+	 */
+	public static function get_template_directory_uri() {
+		return str_replace(get_template(), self::get_template(), get_template_directory_uri());
+	}
+}
